packages/db/src/controller/level.ts

```diff
@@ -175,6 +177,17 @@ export class LevelDbController implements IDatabaseController<Uint8Array, Uint8A
 
     this.metrics?.dbWriteItems.inc({bucket}, itemsRead);
   }
+
+  /** Add collect function to capture metric for db size */
+  private addDbSizeMetricCollector(metrics: ILevelDbControllerMetrics): void {
+    const minKey = Buffer.from([0x00]);
+    const maxKey = Buffer.from([0xff]);
+
+    metrics.dbSizeTotal.addCollect(async () => {
```

Comment:
@wemeetagain I am a bit concerned about the performance impact of this, since the collect function is called every time metrics are fetched. I think for db size we don't need real-time values and could definitely have more delay we could interally limit this to once every 5-10 minutes maybe but I have not tested this against a bigger database size (20-30GB) so I can't really tell, do we have a quick way to run this branch on a existing system? My testing setup is not ready at the moment

Reply:
Yeah, I was hesitant on this PR since it introduced async collection functions when we had none before. (this might have some annoying side effects of making it harder to diagnose event loop lag, since in the past, we've used latency in collecting metrics as a way to measure event loop lag.)

Like you said, you could create a heartbeat function that get gets registered on start, stopped on close, that caches the approx size.

Definitely a way to test this, let me know when you want to test it

Reply:
> this might have some annoying side effects of making it harder to diagnose event loop lag, since in the past, we've used latency in collecting metrics as a way to measure event loop lag

thanks for mentioning this, I think we should not introduce async collection then just for this one use case. I think it is better then to use a interval to collect the metrics as you mentioned

> Definitely a way to test this, let me know when you want to test it

Could maybe test it in the current state, we already track the prom scrape duration and could use that as a indicator

Reply:
> that get gets registered on start, stopped on close, that caches the approx size.

@wemeetagain do you know why the `start` and `stop` is not explicitly called in case of the validator LevelDB? We just instantiate it and implicitly open it by calling `db.get`, maybe it is by design that we lazy open the db but I did not find any explanation for this

https://github.com/ChainSafe/lodestar/blob/6793305c50255b0530438d0a2cb1960d96a3be3b/packages/cli/src/cmds/validator/handler.ts#L101-L104

Reply:
its not intentional, just an oversight

Reply:
fixed in https://github.com/ChainSafe/lodestar/pull/5106, we can now rely on start/stop hook to manage heartbeat function to collect db size metric